File: 123_Best_Time_to_Buy_and_Sell_Stock_III.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if (n < 2) return 0;
        vector<int> former(n, 0);
        vector<int> later(n, 0);
        for (int i = 1, valley = prices[0]; i < n; i++) {
            valley = min(valley, prices[i]);
            former[i] = max(former[i - 1], prices[i] - valley);
        }
        for (int i = n - 2, peak = prices[n - 1]; i > 0; i--) {
            peak = max(peak, prices[i]);
            later[i] = max(later[i + 1], peak - prices[i]);
        }
        int max_profit = 0;
        for (int i = 0; i < n; i++) {
            max_profit = max(max_profit, former[i] + later[i]);
        }
        return max_profit;
    }
};

File: 123_Best_Time_to_Buy_and_Sell_Stock_III.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        '''
        dp[k][i] = max(dp[k][i - 1], p[i] - p[j] + dp[k - 1][j - 1])
        '''
        
        n = len(prices)
        if n < 2: return 0
        dp = [[0] * n for _ in range(3)]
        for k in range(1, 3):
            minima = prices[0]
            for i in range(1, n):
                minima = min(minima, prices[i] - dp[k - 1][i - 1])
                dp[k][i] = max(dp[k][i - 1], prices[i] - minima)
        return dp[-1][-1]


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        '''
        For each iteration, buy the share with the price as low as we can,
        and sell the share with the price as high as we can; then integrate
        the profit of the first trade into the cost of the second buy,
        eventually, the second sell would be the profit of the two transaction.
        '''
        
        buy1 = buy2 = float('inf')
        sell1 = sell2 = 0
        for p in prices:
            buy1 = min(buy1, p)
            sell1 = max(sell1, p - buy1)
            buy2 = min(buy2, p - sell1)
            sell2 = max(sell2, p - buy2)
        return sell2
